### PR TITLE
Add `[@@deriving to_enum]` to generate `val to_enum : t -> int`

### PR DESCRIPTION
**Pull request by mrmr1993**
This PR adds a ppx to enumerate variants not supported by `[@@deriving enum]`. In particular, this supports

```ocaml
type t = A | B [@@deriving to_enum]
type u = A of int | B of bool [@@deriving to_enum]
type v = u [@@deriving to_enum]
type w = [`A | `B] [@@deriving to_enum]
type x = [`A of int | `B of bool] [@@deriving to_enum]
```

Checklist:

- [ ] Document code purpose, how to use it
  + Mention expected invariants, implicit constraints
- [ ] Tests were added for the new behavior
  + Document test purpose, significance of failures
  + Test names should reflect their purpose
- [ ] All tests pass (CI will check this if you didn't)
- [ ] Serialized types are in stable-versioned modules
- [ ] Does this close issues? List them:

### CODE REVIEW

**Comment by psteckler** [comment]
It looks like vanilla `enum` deriver starts the numbering at 0, and not 1. For this program:
```ocaml
type t = [`A of int | `B] [@@deriving to_enum]
type t_ = A | B [@@deriving enum]
```
the generated output is:
```ocaml
type t = [ `A of int  | `B ][@@deriving to_enum]
let (to_enum : t -> int) = function | `A _ -> 1 | `B -> 2
type t_ =
  | A [@@deriving enum]
let min_t_ = 0
and max_t_ = 1
and t__to_enum = function | B -> 1 | A -> 0
and t__of_enum = function | 0 -> Some A | _ -> None
```
Do we want to include the min and max, to complete the emulation?

File: src/lib/ppx_coda/ppx_to_enum/ppx_to_enum.ml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                       ppat_construct ~loc (mk_lid constr.pcd_name)
+                         ( match constr.pcd_args with
+                         | Pcstr_tuple [] ->
+                             None
+                         | _ ->
+                             Some (ppat_any ~loc) )
+                   ; pc_guard= None
+                   ; pc_rhs=
+                       pexp_constant ~loc
+                         (Pconst_integer (Int.to_string (i + 1), None)) } ))]]
```
Comment (psteckler) [inline src/lib/ppx_coda/ppx_to_enum/ppx_to_enum.ml:45]: think you can use `eint` here

File: src/lib/ppx_coda/ppx_to_enum/ppx_to_enum.ml [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                                 (Some (ppat_any ~loc))
+                           | true, _ :: _ ->
+                               (* [`A | `A of int] *)
+                               ppat_or ~loc
+                                 (ppat_variant ~loc label.txt None)
+                                 (ppat_variant ~loc label.txt
+                                    (Some (ppat_any ~loc))) )
+                       ; pc_guard= None
+                       ; pc_rhs=
+                           pexp_constant ~loc
```
Comment (psteckler) [inline src/lib/ppx_coda/ppx_to_enum/ppx_to_enum.ml:86]: `eint`

**Comment by mrmr1993** [comment]
> It looks like vanilla `enum` deriver starts the numbering at 0, and not 1.

Thanks, I had misremembered and didn't check! Should be fixed now.

> Do we want to include the min and max, to complete the emulation?

Good call, added.

**Pull request merged**